modules/nsxt-distributed-firewall-policy/main.tf

```diff
@@ -30,21 +30,47 @@ locals {
       ]
     )
   )
+  source_groups = toset(
+    flatten(
+      [
+        for each_rule in(var.rules != null ? var.rules : []) :
+        [
+          for each_sg in each_rule["source_groups"] : each_sg
+          if try(cidrnetmask(each_sg), null) == null
+          && try(regex(local.ip_address_regex, each_sg), null) == null
+          && try(regex(local.ip_address_range_regex, each_sg), null) == null
+        ]
+      ]
+    )
+  )
   rules = [
     for each_rule in(var.rules != null ? var.rules : []) :
     merge(
       each_rule,
       { destination_groups = [
         for each_dg in each_rule["destination_groups"] :
-        contains(local.sgs, each_dg) ? data.nsxt_policy_group.group[each_dg].path :
+        contains(local.destination_groups, each_dg) ? data.nsxt_policy_group.destination_groups[each_dg].path :
         each_dg
-    ] })
+        ]
+      },
+      { source_groups = [
+        for each_sg in each_rule["source_groups"] :
+        contains(local.source_groups, each_sg) ? data.nsxt_policy_group.source_groups[each_sg].path :
+        each_sg
+        ]
+      }
+    )
   ]
 }
 
 # Lookup existing policy group
-data "nsxt_policy_group" "group" {
-  for_each     = local.sgs
+data "nsxt_policy_group" "destination_groups" {
```

Comment:
@KonradSchieban : you can just use one `data.nsxt_policy_group` and merge `local.destination_groups` & `local.source_groups` using [concat](https://developer.hashicorp.com/terraform/language/functions/concat)  method and use it in `local.rules`

Reply:
but why would we concatenate them if they are used in different parts of the rule? If we concat we would get one array `data.nsxt_policy_group.source_and_destination_groups` but we need to differentiate where to add them in the rule.

Reply:
ok, updated it with `for_each     = setunion(local.destination_groups, local.source_groups)`